0101-nix formatting.md

```diff
@@ -0,0 +1,672 @@
+      c = "cherry";
+      d = "date";
+      e = "elderberry";
+    })
+    feijoa
+    grape
+    honeydew
+    ;
+}
+```
+
+- Technically this is inconsistent because the attribute set is sometimes on the same line as the `inherit` and other times not. However, this is a really niche use case and generally you won't need to write such code anyways so we are willing to sacrifice consistency for a better default.
+
+### If-Then-Else
+
+✅ Good:
+
+```nix
+if predicate then
+  foo
+else
+  bar
+```
+
+- "Pulling up" the `then` keyword is consistent with other similar syntax constructs
+- The bodies are always indented
+
+❌ Bad (old):
+
+```nix
+if predicate
+then foo
+else bar
+```
+
+- TODO explain why bad
+- ~~The keyword at the beginning of the line
+  states clearly the meaning of the content that follows.~~
+- ~~Produces a clean diff when you add more code.
+  For example: adding content to the `else`
+  only produces a diff in the `else`.~~
+
+❌ Bad:
+
+```nix
+if predicate then foo else bar
+```
+
+- One-liners are hard to understand,
+  specially when nested,
+  or when logic gets long.
+- Adding content produces a diff in the entire `if-then-else`.
+
+✅ Good:
+
+```nix
+if something <= 2.0 then
+  if somethingElse then
+    foo
+  else
+    bar
+else if something <= 4.0 then
+  {
+    foo = 10;
+  }
+else if something <= 6.0 then
+  [
+    10
+    20
+  ]
+else if something <= 7.0 then
+  let
+    …
+  in
+    …
+else
+  bar
+```
+
+- It's easy to follow that there are many conditionals.
+- The indentation makes it easy to read
+  which expression is associated to each conditional.
+- Adding or modifying the branches produces a clean diff.
+
+❌ Bad:
+
+```nix
+if cond
+then if
+  looooooooooooooooooooooooooooooooooooooooooooooooooooong
+then foo
+else bar
+else if cond
+then foo
+else bar
+```
+
+- It's complex to distinguish the parent `if-then-else`
+  from the child `if-then-else`
+
+❌ Bad:
+
+```nix
+if cond1 then [
+  foo
+] else if cond2 then
+  bar
+else if cond3 then {
+  baz = "…";
+} else if cond4 then let
+  qux = "…";
+in 
+  qux
+else
+  true
+```
+
+- Inconsistent start of line
+- if/else keywords not aligned: the first `else` is indented like if it was nested
+- Easy to miss the opening brace/bracket, especially when the closing one is far below
+- No clear distinction whether items are in a list or not
+
+✅ Good:
+
+```nix
+if
+  multiline
+    && condition
+then
+  content
+else if
+  myPredicate [
+    multi
+    line
+  ]
+then
+  other content
+else
+  more
+```
+
+- If the condition does not fit one line, expand with indentation, and put the `then` keyword on a new line
+- Recommendation: avoid long conditions
+
+### Lists
+
+✅ Good:
+```nix
+[ foo bar baz ]
+[
+  foo
+  bar
+  baz
+]
+[ ]
+```
+
+- For consistency with other formatting rules, more spacing is preferred
+
+❌ Bad:
+```nix
+[foo bar baz]
+[]
+```
+
+### Function applications
+
+✅ Good:
+```nix
+fun arg1 arg2 arg3
+
+
+callPackage ./some/path {
```

Comment:
Why are curly braces of sets absorbed here but not in if else?

Reply:
We special-cased the last argument of a function call, because it's extremely common

Reply:
That's not an explanation. Having conditionals return one set or another set is also common and could just as well have inline braces:

``nix
if something <= 2.0 then
  if somethingElse then
    foo
  else
    bar
else if something <= 4.0 then {
  foo = 10;
} else if something <= 6.0 then [
  10
  20
]
```

Reply:
Ah right but I see, the drawbacks of absorbing in conditionals are not relevant in function calls.